larpandoradlcontent/LArTwoDReco/DLThreeDClusterSplittingAlgorithm.cc

```diff
@@ -0,0 +1,1350 @@
+            this->GetDistanceToSecVertex(pPathCaloHit, viewSecVtx));
+        features.at("EventHitSep").m_sequence.push_back(
+            this->GetDistanceToEventHit(kdTree, pPathCaloHit, clusterPathHits));
+        features.at("ClusterHitSep").m_sequence.push_back(i == (clusterPath.size() - 1) ? -1.f :
+            this->GetDistanceToClusterHit(clusterHit, clusterPath.at(i + 1)));
+        features.at("GapSep").m_sequence.push_back(i == 0 ? 1.f :
+            this->IsInSameVolume(clusterPath.at(i -1).m_pHit, pPathCaloHit));
+    }
+
+    // Smooth and normalise
+    for (auto &entry : features)
+    {
+        if ((entry.first == "GapSep") || (entry.first == "Longitudinal"))
+            continue;
+
+        entry.second.Smooth();
+        entry.second.Normalise();
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+KalmanFilter2D DLThreeDClusterSplittingAlgorithm::InitialiseKalmanFilter(const ClusterPath &clusterPath) const
+{
+    // Kalman Config
+    const LArTPC *const pTPC(this->GetPandora().GetGeometry()->GetLArTPCMap().begin()->second);
+    const HitType view(clusterPath.begin()->m_pHit->GetHitType());
+    const float pitch(view == TPC_VIEW_U ? pTPC->GetWirePitchU() : view == TPC_VIEW_V ? pTPC->GetWirePitchV() : pTPC->GetWirePitchW());
+    const float processVariance{m_kalmanProcessVarCoeff * pitch * pitch};
+    const float measurementVariance{m_kalmanMeasurementVarCoeff * pitch * pitch};
+
+    // Initialise Kalman fit
+    Eigen::VectorXd init(2);
+    const float seedL(clusterPath.begin()->m_l), seedT(clusterPath.begin()->m_t);
+    init << seedL, seedT;
+    KalmanFilter2D kalmanFilter2D(m_kalmanDelta, processVariance, measurementVariance, init);
+
+    return kalmanFilter2D;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void DLThreeDClusterSplittingAlgorithm::GetFilteredViewSecVertices(const TwoDSlidingFitResult &clusterFit, CartesianPointVector &viewSecVtx) const
+{
+    if (!m_pSecVertexList)
+        return;
+
+    const HitType hitType(LArClusterHelper::GetClusterHitType(clusterFit.GetCluster()));
+    float clusterL1(-1.f), clusterL2(-1.f), clusterT1(-1.f), clusterT2(-1.f);
+    clusterFit.GetLocalPosition(clusterFit.GetGlobalMinLayerPosition(), clusterL1, clusterT1);
+    clusterFit.GetLocalPosition(clusterFit.GetGlobalMaxLayerPosition(), clusterL2, clusterT2);
+    float clusterMinL(std::min(clusterL1, clusterL2)), clusterMaxL(std::max(clusterL1, clusterL2));
+
+    for (const Vertex *const pSecVertex : *m_pSecVertexList)
+    {
+        const CartesianVector pos(LArGeometryHelper::ProjectPosition(this->GetPandora(), pSecVertex->GetPosition(), hitType));
+
+        float thisL(-1.f), thisT(-1.f);
+        clusterFit.GetLocalPosition(pos, thisL, thisT);
+
+        if ((thisL < clusterMinL) || (thisL > clusterMaxL))
```

Comment:
Is a check against thisT useful here as well? Could this not find something that projects onto the fit longitudinally, but that is actually far away in transverse coordinates?

Reply:
Good question. I think I would want to avoid a cut on m_t. This is just meant to avoid cases where we have two hits in the same l coord - the physical translation is having something like a delta ray coming off of the muon track (in the alg I just want to follow the muon track). I think that it's unlikely that a single random far away hit is included in the track? and if a clump of far away hits are included, then I think that the running fit axis would move onto them and m_t would become small anyhow. Do you see what I mean?